File: Weak4/6_subset_sum_equal_to_k.py
from os import *
from sys import *
from collections import *
from math import *

def f(ind, target, arr, dp):
    if target == 0:
        return True
        
    if ind == 0:
        return arr[0] == target
    
    if dp[ind][target] != -1:
        return dp[ind][target]
    
    notTake = f(ind - 1, target, arr, dp)
    take = False

    if arr[ind] <= target:
        take = f(ind - 1, target - arr[ind], arr, dp)
    
    dp[ind][target] = take or notTake
    return dp[ind][target]

def subsetSumToK(n, k, arr):
    dp = [[-1 for _ in range(k + 1)] for _ in range(n)]

    return f(n - 1, k, arr, dp)

File: Weak4/6_subset_sum_equal_to_k.cpp
#include <bits/stdc++.h>
using namespace std;

bool f(int ind, int k, vector<int> &arr, vector<vector<int>> &dp) {
    if (k == 0) return true;
    if (ind == 0) return arr[0] == k;

    if (dp[ind][k] != -1) return dp[ind][k];

    bool notTake = f(ind - 1, k, arr, dp);
    bool take = false;
    if (arr[ind] <= k) {
        take = f(ind - 1, k - arr[ind], arr, dp);
    }

    dp[ind][k] = take || notTake;
    return dp[ind][k];
}

bool subsetSumToK(int n, int k, vector<int> &arr) {
    vector<vector<int>> dp(n, vector<int>(k + 1, -1));
    return f(n - 1, k, arr, dp);
}
